libnmstate/state.py

```diff
@@ -30,6 +30,43 @@
 from libnmstate.prettystate import format_desired_current_state_diff
 from libnmstate.schema import Interface
 from libnmstate.schema import InterfaceType
+from libnmstate.schema import Route
+
+
+class _Route(object):
+    def __init__(self, route):
+        self.state = copy.deepcopy(route)
+        if route.get(Route.STATE) != Route.ABSENT:
+            if Route.TABLE_ID not in route:
+                self.state[Route.TABLE_ID] = Route.USE_DEFAULT_ROUTE_TABLE
+            if Route.METRIC not in route:
+                self.state[Route.METRIC] = Route.USE_DEFAULT_METRIC
+            if Route.NEXT_HOP_ADDRESS not in route:
+                self.state[Route.NEXT_HOP_ADDRESS] = ''
+
+    def __hash__(self):
+        return hash((self.state.get(Route.STATE),
```

Comment:
I think this shows why saving a dict as the object state is improper here.
You need to re-access the value for each key, on every call to `__hash__`, instead of just directly specifying the attributes (and extracting the values once in the constructor).

You have an example here of something similar: https://github.com/nmstate/nmstate/blob/master/libnmstate/nm/wired.py#L27

Reply:
The `class State` will use that dict for future handling. 
Saving them as object property will force me to do `to_dict()` in `State.merge_config()`.

Reply:
If this is the deal breaker, let me know I can do extra work to make this PR merged.

Reply:
I consider the encapsulated data to server the class and not to be transparent to the user, especially when it is "complex". If the inner usage needs to access the properties, it makes it reasonable to expose them as attributes internally and do the translation for the input and output to whatever format is needed.
If you look at this object now, without the usage, you have hard time explaining why you keep it as a dict internally, which is a smell that needs attention.